main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,379 @@
+
+                        // Unlock the selectLock
+                        unlockLock(selectLock);
+
+                        None
+                    } else {
+                        // We have to wait for an element
+                        // Add our condition to all channels to get notified when a new element is added
+                        let addGetterToChannel = c -> {
+                            let pair = (selectLock, selectCondition);
+                            let Mpmc(_, _, _, _, _, waitingGetters, _) = c;
+
+                            // Add LockConditionPair to the channel
+                            MutList.push!(pair, waitingGetters)
+                        };
+                        Array.foreach(addGetterToChannel, channels);
+
+                        // Unlock all channels in sorted order, so other threads may input elements
+                        Array.foreach(unlockChannel, sortedChannels);
+
+                        // Wait for an element to be added to any of the channels
+                        awaitCondition(selectCondition);
+
+                        // Unlock the selectLock
+                        unlockLock(selectLock);
+
+                        // try again
+                        selectHelper(channels, sortedChannels, selectLock, selectCondition, hasDefault)
+                    }
+                }
+            }
+        } else {
+            internalBug("selectHelper", "thread interrupted")
+        }
+
+    ///
+    /// Retrieves the element of the first channel in the array that has one, if such a channel exists.
+    ///
+    def firstAvailableElement(index: Int, channels: Array[ChannelImpl]): Option[(Int, Boxed)] & Impure =
+        if (index < 0 or index >= Array.length(channels)) {
+            // Passed `index` out of bounds
+            internalBug("firstAvailableElement", "out of bounds ${index}, length ${Array.length(channels)}")
+        } else {
+            match tryGet(channels[index]) {
+                case None =>
+                    // try the next channel
+                    if (index == Array.length(channels)-1) None
+                    else firstAvailableElement(index + 1, channels)
+                case Some(element) => {
+                    // There is a waiting element in this channel.
+                    // Return the element and the index of this channel
+                    Some((index, element))
+                }
+            }
+        }
+
+
+    // ----- Helper Methods -----
+
+    ///
+    /// `bug!(...)` with a formatted error of filename, `method`, and `description` that blames the implementation devs.
+    ///
+    def internalBug(methodName: String, description: String): a =
+        bug!("implementation error in ChannelImpl.flix.${methodName}, ${description}")
+
+    ///
+    /// `bug!(...)` with a formatted error of filename, `method`, and `description` that blames the input.
+    ///
+    def inputBug(methodName: String, description: String): a =
```

Comment:
Inline, its only used in one place. Besides it should be the job of bug to report the line number (even if it does not currently).